src/pagination.rs

```diff
@@ -0,0 +1,312 @@
+        // list of available indexes, in `tags` like cases
+        if let Some(indexes) = self.indexes {
+            object.insert(
+                "indexes".into(),
+                liquid::value::Value::Array(
+                    indexes.iter().map(liquid::value::Value::scalar).collect(),
+                ),
+            );
+        }
+        object.insert(
+            "index".into(),
+            liquid::value::Value::scalar(self.index as i32),
+        );
+        object.insert(
+            "index_permalink".into(),
+            liquid::value::Value::scalar(self.index_permalink),
+        );
+        if let Some(previous_index_permalink) = self.previous_index_permalink {
+            object.insert(
+                "previous_index".into(),
+                liquid::value::Value::scalar(self.previous_index as i32),
+            );
+            object.insert(
+                "previous_index_permalink".into(),
+                liquid::value::Value::scalar(previous_index_permalink),
+            );
+        }
+        if let Some(next_index_permalink) = self.next_index_permalink {
+            object.insert(
+                "next_index".into(),
+                liquid::value::Value::scalar(self.next_index as i32),
+            );
+            object.insert(
+                "next_index_permalink".into(),
+                liquid::value::Value::scalar(next_index_permalink),
+            );
+        }
+        object.insert(
+            "first_index_permalink".into(),
+            liquid::value::Value::scalar(self.first_index_permalink),
+        );
+        object.insert(
+            "last_index_permalink".into(),
+            liquid::value::Value::scalar(self.last_index_permalink),
+        );
+        object.insert(
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc.front.pagination.as_ref().unwrap();
+    let mut all_posts = posts_data.to_vec();
+    match config.include {
+        Include::All => {
+            sort_posts(&mut all_posts, &config);
+            create_all_paginators(&mut all_posts, &doc, &config)
+        }
+        Include::None => Ok(Vec::new()), // user can set `include: None`
+    }
+}
+
+fn extract_value(a: &liquid::value::Value, key: &str) -> Option<liquid::value::Scalar> {
+    a.get(&key.into())
+        .and_then(|sort_key| sort_key.as_scalar().cloned())
+}
+
+// sort posts by multiple criteria
+fn sort_posts(posts: &mut Vec<liquid::value::Value>, config: &PaginationConfig) {
+    // Boxing needed to compile, don't really understand why
+    let order: Box<Fn(liquid::value::Scalar, liquid::value::Scalar) -> Ordering> = match config
```

Comment:
Each closure has a unique, compiler-defined type.  What you are doing is forcing the unnameable structs into being trait objects.

Alternatively, you should be able to decay these into function pointers (`fn(liquid::value::Scalar, liquid::value::Scalar) -> Ordering`, emphasis on the lower-case `fn`) and avoid the boxing entirely.

Reply:
I didn't knew about function pointers before this :D Thanks!